src/kalai/pass/kalai/a_annotate_ast.clj

```diff
@@ -49,42 +52,65 @@
     #'ref
     #'agent})
 
-(defn ast-t [ast]
-  (m/rewrite ast
-    ;; (atom x)
-    {:op   :invoke
-     :fn   {:var (m/pred ref-vars)}
-     :args [?value . _ ...]}
-    ~(ast-t ?value)
-
-    ;; ^{:t :long, :tag Long} x
-    {:op   :with-meta
-     :meta {:form {:t ?t :tag ?tag}}
-     :expr ?expr}
-    ~(or ?t
-         (get types/java-types ?tag)
-         (ast-t ?expr))
-
-    {:op   :local
-     :form ?form}
-    ~(resolve-t ?form ast)
-
-    ;; Last resort: 1
-    {:op  :const
-     :val ?val}
-    ~(get types/java-types (type ?val))
-
-    ;; Last resort: Clojure type inferred
-    {:o-tag (m/pred some? ?o-tag)}
-    ~(or (get types/java-types ?o-tag)
-         ;; TODO: breaking the rules for interop...
-         ;; is this a bad idea?
-         ;; note that some o-tags are unhelpful,
-         ;; like PersistentMap
-         (#{StringBuffer} ?o-tag))
-
-    ?else
-    nil))
+(defn ast-t
+  "Return the type represented by an AST node"
+  ([ast]
+   (ast-t ast ast))
+  ([ast root]
+   (m/rewrite ast
+     ;; (atom x)
+     {:op   :invoke
+      :fn   {:var (m/pred ref-vars)}
+      :args [?value . _ ...]}
+     ~(ast-t ?value ast)
+
+     ;; ^{:t :long, :tag Long} x
+     {:op   :with-meta
+      :meta {:form {:t ?t :tag ?tag}}
+      :expr ?expr}
+     ~(or ?t
+          (get types/java-types ?tag)
+          (ast-t ?expr ast))
+
+     ;; Following the type inference precedence order, if we don't see type
+     ;; info on the binding name itself, then check the initial value.
+     ;; This enables the initial-value-to-binding propagation of type metadata.
+     {:op   :binding
+      :form (m/and
+              (m/app #(resolve-t % root) ?t)
+              (m/guard (some? ?t)))}
```

Comment:
We can replace
`(m/and (m/app #(resolve-t % root) ?t) (m/guard (some? ?t)))`
with
`(m/app #(resolve-t % root) (m/pred some? ?t))`

Reply:
Done.